### ISSUE TITLE
update CRAN version of r4ss

### ISSUE THREAD

**Issue by iantaylor-NOAA** [r4ss/r4ss#206]
Current version on CRAN is out of date and doesn't work for many 3.30 models.

EDIT 10 May 2022: checklist added from comments below:
- [x] re-read http://r-pkgs.had.co.nz/release.html#release-submission
- [x] update `NEWS.md`
- [x] update `cran-comments.md`
- [x] temporarily add the following to `.Rbuildignore` 
  - [x] `inst/extdata` to avoid NOTE about package size in package check 
  - [x] `tests` because it depends on the example model files in `inst/extdata`
  - [x] `vignettes` because it depends on the example model files in `inst/extdata`
- [x] use `devtools::release()` which asks for confirmation of the following checks
  - [x] `devtools::spell_check()` # will be faster now that `inst/WORDLIST has been created
  - [x] `devtools::check()`
  - [x] `devtools::check_rhub()`
  - [x] `devtools::check_win_devel()` # identified some issues with broken URLs
  - [x] `revdepcheck::revdep_check()` # checks ss3sim and SSMSE
- [x] confirm the submission to CRAN via link in automated email after uploading
- [x] merge `CRAN` branch (if used for submission)
- [x] tag the code with the version number

**Comment by iantaylor-NOAA** [comment]
Hoping to update the CRAN version this week. If it goes well, that might become a more regular occurrence as there have been frequent reports of folks having trouble installing via `devtools::install_github()`.

**Comment by iantaylor-NOAA** [comment]
The following should be done before submission
- [ ] temporarily add `tests` and `inst/extdata` to `.Rbuildignore` to avoid NOTE about package size in package check
- [ ] update `cran-comments.md`
- [ ] re-read http://r-pkgs.had.co.nz/release.html#release-submission
- [ ] use `devtools::release()`

**Comment by iantaylor-NOAA** [comment]
Submitting to CRAN "this week" as I suggested last week didn't work out. Will attempt to make submission next week after consultation with @kellijohnson-NOAA about whether to include changes in Pull Request #341 along with those in Pull Request #340 in the submission.

**Comment by kellijohnson-NOAA** [comment]
@iantaylor-NOAA perhaps we can just use this issue to track submitting to CRAN rather than opening a new one. I tried `rhub::check()` last night but it failed because the tests took too long. I notice now you mentioned adding the tests directory to .Rbuildignore but I am wondering if we should just mark some tests to not run on CRAN rather than ignore the folder?
```
   Build timed out (after 30 minutes). Marking the build as failed.
   Build was aborted
   Pinging https://builder.r-hub.io/build/FAILURE/r4ss_1.43.2.tar.gz-53d872c0104d47d9841d3a673c0af5d6/2022-05-10T01:38:17Z
   {"status":"ok"}
   Finished: FAILURE
```
I also got an email with the following summary about folder sizes. I know that we are aware that extdata needs to be removed but it looks like the number of files/size of some files in .\R\ is also a problem.
```
NOTES:
* checking installed package size ... NOTE
  installed size is 19.4Mb
  sub-directories of 1Mb or more:
    R         1.6Mb
    extdata  17.0Mb
```

**Comment by iantaylor-NOAA** [comment]
Thanks for doing these checks, @kellijohnson-NOAA.

The CRAN policy (https://cran.r-project.org/web/packages/policies.html#:~:text=Packages%20should%20be%20of,a%20maximum%20of%205MB.) says the size limit is 5MB with no mention of limits for sub-directories, so I would just ignore the note about the R directory.

Regarding the files in `extdata` used in the tests, the files for the simple_3.30.13 model add up to 4.5MB. I think the only tests that we could keep without big model output files would be those related to read/write of the input files. Testing old model files isn't as important as recent ones anyway, so the tests should be revised, and part of that revision could include a streamlined subset for CRAN. But I'm not sure it makes sense to hold up the CRAN submission to wait for revising the tests.

Lastly, I moved the checklist to the first comment so that it appears in the progress bar on the issues list. I've never seen github documentation for that requirement, but it has been a consistent pattern. I previously also used a project checklist https://github.com/r4ss/r4ss/projects/1#card-41803198, but I'm not sure that works as well.

**Comment by iantaylor-NOAA** [comment]
@kellijohnson-NOAA and @k-doering-NOAA,
I ran `revdepcheck::revdep_check()` and it found the error associated with {ss3sim}.
I assume that making a change in ss3sim on github won't help because it's checking the CRAN version.
Any suggestions on whether we (1) ignore it assuming that CRAN's checks aren't this thorough, or (2) something else?

```
    Running examples in 'ss3sim-Ex.R' failed
    The error most likely occurred in:
    
    > ### Name: change_data
    > ### Title: Change the data that is available as output from an SS operating
    > ###   model.
    > ### Aliases: change_data
    > 
    > ### ** Examples
    > 
    ...
    > years <- c(5, 10, 15)
    > types <- c("len", "age")
    > file_in <- r4ss::SS_readdat(file.path(d, "models", "cod-om", "codOM.dat"),
    +   version = NULL, verbose = FALSE)
    Warning: The `version` argument of `SS_readctl()` must be 3.24 or 3.30 as of r4ss 1.43.2.
    This warning is displayed once every 8 hours.
    Call `lifecycle::last_lifecycle_warnings()` to see where this warning was generated.
    Error in if (nver < 3) { : argument is of length zero
    Calls: <Anonymous>
    Execution halted
 ```

**Comment by yukio-takeuchi** [comment]
It looks, the error is caused assuming that version is numeric or character representing numeric. On the other hand one block above where a warning produced allow NULL for version.

I suggest to change warning to error or for NULL value of version

**Comment by k-doering-NOAA** [comment]
Hmm, I'm not really sure, it depends what CRAN does with a failed dependency check? If it is possible to get r4ss on CRAN with it failing? If not, then maybe we will to do as @yukio-takeuchi suggests and allow for NULL to work for 1 more version of r4ss (because the next time ss3sim is updated on CRAN, it will no longer use NULL for version)

**Comment by kellijohnson-NOAA** [comment]
I think this is what cran-comments.md is supposed to be used for. Regardless, I do not know if the check down-stream dependencies. [Here is an example cran-comments.md](https://github.com/rcannood/GillespieSSA/blob/master/cran-comments.md) that included the results from the reverse dependencies check and we could just add a note that {ss3sim} will be submitted shortly.

**Comment by iantaylor-NOAA** [comment]
Thanks for all the feedback and sorry for the slow response. 

The R packages book (https://r-pkgs.org/release.html#release-deps) says "it’s your responsibility to check that downstream dependencies" and that if there's a failure, "you should give package authors at least two weeks to fix the problem before you submit your package to CRAN".

The `revdepcheck::revdep_check()` conveniently wrote this .md file to paste to the bottom of cran-comments.md
```
## revdepcheck results

We checked 2 reverse dependencies, comparing R CMD check results across CRAN and dev versions of this package.

 * We saw 1 new problems
 * We failed to check 0 packages

Issues with CRAN packages are summarised below.

### New problems
(This reports the first line of each new failure)

* ss3sim
  checking examples ... ERROR
```
So if you don't include those results you probably come across as lazy. And I'd rather not put pressure on ourselves with an explanation about a new version of {ss3sim} coming soon.

So in commit 0459f74, I've just pushed back the lifecycle warning and automatically converted NULL to "3.30". We can revert or revise that change in the github version of {r4ss} at any time after the CRAN submission. This allows the {revdepcheck} to pass with "We saw 0 new problems" as reflected in the revised cran-comments.md (841648a).

Please comment if you have any concerns. If there are no objections or further revisions, I will submit to CRAN sometime tomorrow.

**Comment by kellijohnson-NOAA** [comment]
Thanks Ian. I thought your fix was pretty brilliant actually because it doesn't require a future change in {r4ss}. Thank you for submitting this.

**Comment by kellijohnson-NOAA** [comment]
I am guessing that the line number 53 pertains to the rendered DESCRIPTION not the submitted file which has markdown.

**Comment by kellijohnson-NOAA** [comment]
Actually, I think it might be the raw file and r4ss on line 51. maybe if you use {r4ss} instead of just r4ss?

**Comment by iantaylor-NOAA** [comment]
Success with the resubmission:

>Dear maintainer,
>
>thanks, package r4ss_1.44.0.tar.gz is on its way to CRAN.
>
>Best regards,
>CRAN teams' auto-check service
>Package check result: OK
>
>No changes to worse in reverse depends.

I learned a few things along the way:
* CRAN does run reverse dependency checks, so it's good we fixed things for {ss3sim}

* It was indeed the occurrence of "r4ss" that raised the spell check error flagging the "ss" part, but it wasn't on line 51 of the raw file. I wasn't getting the NOTE with `devtools::check()` but `devtools::check_win_devel()` helped me confirm that I had fixed the issue before resubmission.
  - the CRAN spell check uses `aspell` and differs from the `devtools::spell_check()` making it not currently worth the effort to add words to a custom dictionary (see discussion here: https://github.com/r-lib/usethis/issues/1466)
  - Instead, non-standard words in the "Description" field of the DESCRIPTION file should have single quotes around them as noted [here](https://cran.r-project.org/doc/manuals/r-release/R-exts.html#:~:text=single%20quotes%20for%20non%2DEnglish%20usage%2C%20including%20names%20of%20other%20packages%20and%20external%20software) in "Writing R Extensions". I chose to change `r4ss` to `'{r4ss}'`. The occurrence of `r4ss` in the "Package" field of DESCRIPTION was not a problem and single quotes there messed things up. 

* I ended up just removing the URLs rather than dig into why some of them were flagged as possibly invalid or complaints about not using `\doi{}`.
  * I'm undecided on the value of adding the URLs back for versions of {r4ss} on github (because they're useful references related to the package functions) or leaving them out until I understand better why they got flagged to facilitate future CRAN submissions.
  * I can't find a single instance of `\doi{}` in R package documentation (in part because search engines don't work with the backslash) and I couldn't get it to work for me (issue at #661). Reading the info at https://cran.r-project.org/doc/manuals/r-release/R-exts.html#index-_005cdoi_007bnumbers_007d, https://ctan.org/pkg/doi, and https://github.com/ho-tex/doi was not helpful.

**Comment by kellijohnson-NOAA** [comment]
@iantaylor-NOAA is there any way in the future that one version of `simple` could be left in extdata for the CRAN submission? {ss3sim} uses this model for some examples and tests, though I have now used `skip_on_cran()` for these instances, I still think it would be good for users to have one version of the model.

**Comment by iantaylor-NOAA** [comment]
@kellijohnson-NOAA, the biggest problem is the covar.sso file, which for simple is about 3 MB (relative to a CRAN limit of 5 MB. We could either leave that file out and skip any tests which depend on that file, or we could make a model with fewer years of data to reduce the file size).

I did leave in place the input files for simple_3.30.13 because they are small and are used by the example for `SS_read()` and not wrapped in `\dontrun{}`.

Let me experiment with shrinking the covar.sso through changes to simple and report back here or in a new issue.

**Comment by iantaylor-NOAA** [comment]
I created a modified version of "simple" in the `[smaller_test_model_206](https://github.com/r4ss/r4ss/compare/smaller_test_model_206)` branch (6d28c52) that is about 2 MB for all the files and I think could work fine on CRAN. Next step would be to work on revising the existing tests to point to this model instead of the various SS versions. That effort deserves a separate github issue but that can wait until I'm back next week.

The messy script to modify the ["Simple" in test-models](https://github.com/nmfs-stock-synthesis/test-models/tree/main/models/Simple) is attached as ([simple_small_changes.R.txt](https://github.com/r4ss/r4ss/files/8914293/simple_small_changes.R.txt)) in case it's needed in the future. I don't see a need to add it to the repo itself but we could do that if it would be useful.

Thanks @kellijohnson-NOAA for the inspiration to tackle this longstanding problem.

**Comment by iantaylor-NOAA** [comment]
The smaller test model that can be used on CRAN is now in place (thanks @k-doering-NOAA for remembering to merge the workflow changes), so I'm closing this issue and opening a placeholder for a future CRAN submission.
